### PR TITLE
Add WithDataVolume to Redis Insights

### LINKED ISSUE #6299: Add WithDataVolume to Redis Insights
## Background and Motivation

When using Redis Insights, every time you F5 and launch the container you get prompted with a privacy and license acceptance prompt the first time you hit the Redis Insights page.

To solve this, we should add a method that configures a data volume on the container. That way you only need to accept the license and privacy statement the first time.

## Proposed API

```diff
namespace Aspire.Hosting;

public static class RedisBuilderExtensions
{
+    public static IResourceBuilder<RedisInsightResource> WithDataVolume(this IResourceBuilder<RedisInsightResource>, string? name = null);
}
```

## Usage Examples

```csharp
var redis = builder.AddRedis("redis")
    .WithDataVolume()
    .WithRedisInsight(i => i.WithDataVolume());
```

### PR DESCRIPTION
**Pull request by Alirexaa**
## Description

This pull request introduces new features to the `Aspire.Hosting.Redis` project, including methods for adding data volumes and bind mounts to Redis Insight resources, along with corresponding tests to ensure data persistence between usages. The most important changes are detailed below:

### New Features:
* Added `WithDataVolume` and `WithDataBindMount` methods to `RedisBuilderExtensions` to support adding named volumes and bind mounts for data folders in Redis Insight container resources. (`src/Aspire.Hosting.Redis/PublicAPI.Unshipped.txt`, `src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs`) [[1]](diffhunk://#diff-47e83ab291b80fd27cd1ca712642e384629f63854cb28845006a6bcb4d090ebfR5-R6) [[2]](diffhunk://#diff-a5074219d9ede6622333dbc0f6da94c74c45ea8c68faa1c2829b57cd3a614659R379-R406)

### Testing Enhancements:
* Introduced a new test `RedisInsightWithDataShouldPersistStateBetweenUsages` in `RedisFunctionalTests` to verify that Redis Insight data persists between container restarts, using either volumes or bind mounts. (`tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs`)
* Added `Newtonsoft.Json.Linq` to `RedisFunctionalTests` for JSON handling in the new test. (`tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs`)

Fixes #6299

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [ ] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6432)

### CODE REVIEW

**Comment by Alirexaa** [comment]
I couldn't check the tests will pass or not because I can't build repo for the following error
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error (active)	NU1100	Unable to resolve 'Microsoft.DeveloperControlPlane.windows-amd64 (= 0.8.11)' for 'net8.0'. PackageSourceMapping is enabled, but the following source(s) were not considered: C:\Program Files\dotnet\library-packs, dotnet-eng, dotnet-public, dotnet9.	Aspire.Hosting.AppHost	D:\Workspace\aspire\src\Aspire.Hosting.AppHost\Aspire.Hosting.AppHost.csproj	1		
```

File: src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs [hunk 1]
```diff
@@ -376,4 +376,32 @@ public static IResourceBuilder<RedisResource> WithPersistence(this IResourceBuil
...
         }), ResourceAnnotationMutationBehavior.Replace);
     }
+
+    /// <summary>
+    /// Adds a named volume for the data folder to a Redis Insight container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "<Pending>")]
```
Comment (eerhardt) [inline src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:386]: > Justification = "<Pending>"

This should be filled out.

File: tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@
 using StackExchange.Redis;
 using Xunit;
 using Xunit.Abstractions;
+using Newtonsoft.Json.Linq;
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs:18]: Can you change this to use System.Text.Json? We shouldn't be adding new code using Newtonsoft.Json.

File: tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs [hunk 3]
```diff
@@ -428,6 +429,171 @@ public async Task PersistenceIsDisabledByDefault()
...
+                analytics = false,
+                notifications = false,
+                encryption = false,
+            }
+        });
+
+        var apiUrl = $"/api/settings";
+
+        var response = await client.PatchAsync(apiUrl, jsonContent, ct)
+                            .ConfigureAwait(false);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs:590]:
```suggestion
```

You don't need this in tests.

File: tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs [hunk 4]
```diff
@@ -428,6 +429,171 @@ public async Task PersistenceIsDisabledByDefault()
...
+        try
+        {
+            using var builder1 = TestDistributedApplicationBuilder.CreateWithTestContainerRegistry(testOutputHelper);
+            IResourceBuilder<RedisInsightResource>? redisInsightBuilder1 = null;
+            var redis1 = builder1.AddRedis("redis")
+                .WithRedisInsight(c => { redisInsightBuilder1 = c; });
+            Assert.NotNull(redisInsightBuilder1);
+
+            if (useVolume)
+            {
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs:452]: Would the test look nicer if we moved the `WithDataVolume` call into the `WithRedisInsight` callback instead? That way we didn't have this variable capturing.

Comment (Alirexaa) [reply tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs]: variable capturing is necessary for another part of the test.
see:
https://github.com/dotnet/aspire/pull/6432/files#diff-3b0aa8b65b60aa9db2aff71ccdede3578734eccad1ca79454241c28b3ecc73b3R556-R569

File: src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs [hunk 5]
```diff
@@ -438,4 +438,32 @@ public static IResourceBuilder<RedisResource> WithPersistence(this IResourceBuil
...
         }), ResourceAnnotationMutationBehavior.Replace);
     }
+
+    /// <summary>
+    /// Adds a named volume for the data folder to a Redis Insight container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "name")]
```
Comment (eerhardt) [inline src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:448]: I think we need a better justification than just "name".

File: tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs [hunk 6]
```diff
@@ -121,111 +121,6 @@ public async Task VerifyRedisResource()
         Assert.Equal("value", value);
     }
 
-    [Fact]
-    [RequiresDocker]
-    public async Task VerifyDatabasesAreNotDuplicatedForPersistentRedisInsightContainer()
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs:126]: Why are you deleting this test?

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**